// element-plus/lib/components/steps/src/steps3.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('../../../constants/index2.js');
require('../../../utils/index2.js');
var props = require('../../../utils/vue/props2.js');
var event = require('../../../constants/event2.js');
var core = require('@vueuse/core');

const stepsProps = props.buildProps({
  space: {
    type: [Number, String],
    default: ""
  },
  active: {
    type: Number,
    default: 0
  },
  direction: {
    type: String,
    default: "horizontal",
    values: ["horizontal", "vertical"]
  },
  alignCenter: {
    type: Boolean
  },
  simple: {
    type: Boolean
  },
  finishStatus: {
    type: String,
    default: "finish",
    values: ["wait", "process", "finish", "error", "success"]
  },
  processStatus: {
    type: String,
    default: "process",
    values: ["wait", "process", "finish", "error", "success"]
  }
});
const stepsEmits = {
  [event.CHANGE_EVENT]: (newVal, oldVal) => core.isNumber(newVal) && core.isNumber(oldVal)
};

exports.stepsEmits = stepsEmits;
exports.stepsProps = stepsProps;
